### ISSUE TITLE
feat: refresh describe resource on change

### ISSUE THREAD

**Issue by mosheavni** [Ramilito/kubectl.nvim#228]
I usually want to follow when I describe an object,
can the float refresh the output?
Thanks.

**Comment by Ramilito** [comment]
I think so, the informer is still active when we have the describe object open. Not sure yet what the best way to implement it is.

**Comment by mosheavni** [comment]
we can start by a view similar to logs that just reloads every time
we could save the result every time and change the view only if something was changed.
using resource builder i don't think it would be too much of a hassle

**Comment by Ramilito** [comment]
That approach would require us to diff the yaml outputs which probably needs a rocks library, I think we can let kubernetes do this for us by hooking into the informer and listen for the modified/deleted events.

We also need to solve the window being recreated when reloading floats, we can't really refresh it at the moment using resourcebuilder.
I have a working PoC here #297 that seems to be working, I still need to test it properly and add some adjustments but it's looking very promising!

**Comment by mosheavni** [comment]
@Ramilito  not working for me.

**Comment by Ramilito** [comment]
Weird, what's the behavior you're getting?
I'm testing it like this:
Go to a pod, hit describe on it.
In another window, I add a label or modify the same pod.
Doesn't the view refresh?

**Comment by mosheavni** [comment]
I want it to refresh also the events, imagepull, container starting, container started, probes succeeded/failed. Are these events sent on the informer?

**Comment by mosheavni** [comment]
Events have a timestamp which updates every second. Is it possible to fetch asynchronous every second and replace the buffer content (as opposed to clearing and re-rendering the entire float)?

**Comment by Ramilito** [comment]
> I want it to refresh also the events, imagepull, container starting, container started, probes succeeded/failed. Are these events sent on the informer?

Yes they should be, is that not happening?

> Events have a timestamp which updates every second. Is it possible to fetch asynchronous every second 

The timestamp is during redraw compared to the value we got on first draw, it's not polling the kubernetes api every tick that changes, just recalculating it with current system time vs the time we did the first request and then we use the informer for any real changes to the state.

> replace the buffer content (as opposed to clearing and re-rendering the entire float)

hmm I'm not sure I see a distinction there, we are only replacing the buffer content and that's what re-rendering the float is, since it's just a buffer. The flickering you see is caused by us clearing the buffer before actually having the new information, we might be able to delay the clearing until we have the content we want to replace it with

**Comment by mosheavni** [comment]
![2024-09-09_09-55-57 (1)](https://github.com/user-attachments/assets/039c2337-4c55-4020-a4fb-a5c4373c438a)

it's actually not refreshing at all.

regarding the re-rendering, your suggestion sounds good. change the buffer only when we get the new describe output and not print the intermediate `Loading...`

**Comment by Ramilito** [comment]
I didn't mean that the timers would update no, only real state changes as in new events would be refreshing.

To get what you mean then we would be needing to do a pull based thing like you suggested but I really don't like that approach, spams the kubernetes api for basically only getting an updated time.

Alternatively bypassing the kubectl describe all together and creating our own view that we can reevaluate on a loop would be better but that's a lot of work...not sure where to go from here on this, do we have any other options?

**Comment by Ramilito** [comment]
That shouldn't be happening, will check it out soon!

**Comment by mosheavni** [comment]
can we reopen meanwhile? @Ramilito

**Comment by mosheavni** [comment]
I ran `kubectl describe pod blabla -v=7`
all it does is fetch the pod's yaml and the pod's events.
i think it's pretty reasonable to create our version of this view

**Comment by mosheavni** [comment]
I checked a bit the kubernetes cli code for the describe, it seems every describe is different:
https://github.com/kubernetes/kubectl/blob/262825a8a665c7cae467dfaa42b63be5a5b8e5a2/pkg/describe/describe.go#L195-L239

and there's a generic describe
I think if we create a class and provide the basic structure for each describe, it won't be too hard to implement our own describe.
Yes it could take some time. I'm willing to start soon the implementation.
@Ramilito  WDYT?

**Comment by Ramilito** [comment]
Give me a day or two to think about it, it's quite the commitment to go that route and maybe there are easier ways

**Comment by mosheavni** [comment]
> Give me a day or two to think about it, it's quite the commitment to go that route and maybe there are easier ways



> Give me a day or two to think about it, it's quite the commitment to go that route and maybe there are easier ways

if we could achieve 3 things, we could steer away from this path:
- refresh the describe float also when an event was added to the resource. this is the API call that is being made when I describe a pod:

```
I0919 18:05:49.507086   66633 round_trippers.go:463] GET https://blankkkk.gr7.us-east-1.eks.amazonaws.com/api/v1/namespaces/production/events?fieldSelector=involvedObject.name%3Dazure-analytics-tsdb-proxy-pass-bla-bla%2CinvolvedObject.namespace%3Dproduction%2CinvolvedObject.uid%3Dhiri-biri-bla-bli-ce500f27dcdd&limit=500
```
so if that also gives a different result, we need to refresh
- remove the `Loading...` message and keep the output until new output is ready to be displayed
- if the cursor was at the last lane, place the cursor at the last line of the newly refreshed data.
else: place the cursor back in that line after a refresh happened.

**Comment by Ramilito** [comment]
> remove the Loading... message and keep the output until new output is ready to be displayed (will be solved by https://github.com/Ramilito/kubectl.nvim/pull/320)
> if the cursor was at the last line, place the cursor in the last line of the newly refreshed data.
> else: place the cursor back in that line after a refresh happened

These are solved, I will try to understand why that event isn't triggering the informer (it should I believe)

**Comment by Ramilito** [comment]
Apparently I was incorrect in that the event should also trigger a watch event on the invlovedObject from that event.
Did a fix for that in #322 and now it looks pretty nice imo, could you test it out and see if you are happy with that as well @mosheavni?

**Comment by mosheavni** [comment]
yes it's working great! took a look on your changes and it's a nice addition of how you handle events, BIG LIKE.
